### PR TITLE
R3f mixin

### PR DESCRIPTION
**Pull request by moyapchen**
Implementation of "Better Fine-Tuning by Reducing Representational Collapse", Aghajanyan et al. (2020). https://arxiv.org/ans/2008.03156."

Same as https://github.com/facebookresearch/ParlAI/pull/3177, but did in a new branch cause that seemed easier.

NOTE: Still need to run evaluations to see end performance; a few settings here (documented) are remnants of that but general structure should be fine. 

Test plan:

GPT2:
Ran

```
parlai train -t taskmaster2 -veps 0.1 --batchsize 2 --model parlai.agents.hugging_face.gpt2:Gpt2Agent_R3F --model-file ~/tmp/modlefile --use-r3f true --r3f-lambda 0.001 --r3f-noise-type normal --r3f-eps 0.001 --domains music --model-file ~/tmp/model --r3f-decoder -noise true --r3f-encoder-noise false
```

and verified output made sense relative to

    parlai train -t taskmaster2 -veps 0.1 --batchsize 2 --model parlai.agents.hugging_face.gpt2:Gpt2Agent --model-file ~/tmp/modlefile

BART:
Ran

    parlai train -t taskmaster2 -veps 0.1 --batchsize 2 --model parlai.agents.bart.bart:BartAgent_R3F --fp16 True --label-truncate 512 --log-every-n-secs 10 -lr 0.0001 --lr-scheduler invsqrt --optimizer adam --save-after-valid True --text-truncate 512 --warmup-updates 100 --fp16-impl apex --update-freq 2 --gradient-clip 1.0 --skip-generation False -vp 50 --max-train-time 171000 -vmt jga -vmm max -dynb full -tblog true --load-from-checkpoint true --use-r3f true --r3f-lambda 0.001 --r3f-noise-type normal --r3f-eps 0.001 --domains music --model-file ~/tmp/model

and made sure output made sense relative to

    parlai train -t taskmaster2 -veps 0.1 --batchsize 1 --model bart --fp16 True --label-truncate 512 --log-every-n-secs 10 -lr 0.0001 --lr-scheduler invsqrt --optimizer adam --save-after-valid True --text-truncate 512 --warmup-updates 100 --fp16-impl apex --update-freq 2 --gradient-clip 1.0 --skip-generation False -vp 50 --max-train-time 171000 -vmt jga -vmm max -dynb full -tblog true --load-from-checkpoint true --domains music --model-file ~/tmp/test

Debug logs still present in the first commit of this PR to show that usage of hooks made sense.

### CODE REVIEW

**Review by stephenroller — COMMENTED** [review]
Quick comments. Overall looks good. Looking forward to collating experimental results.

File: parlai/agents/transformer/module_extensions.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/agents/transformer/module_extensions.py:1]: Was wondering when the "yeah this file could be better named" comment would come in. :)

**Comment by moyapchen** [comment]
Closing/abandoning in favor of https://github.com/facebookresearch/ParlAI/pull/3204
